### PR TITLE
Parsing YAML config to XML leads to incorrect structures

### PR DESCRIPTION
**Pull request by mdonkers**
Changelog category (leave one):

- Bug Fix (user-visible misbehaviour in official stable or prestable release)

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Fixed the incorrect translation YAML config to XML. 

Detailed description / Documentation draft:
Added test to show incorrect behaviour. Parsing YAML to XML leads to repeated parent nodes so that:

```yaml
  networks:
  - ip: "10.1.6.168"
  - ip: "::1"
  - ip: "127.0.0.1"
```
generates
```xml
 <networks>
 <ip>10.1.6.168</ip>
 </networks>
 <networks>
 <ip>::1</ip>
 </networks>
 <networks>
 <ip>127.0.0.1</ip>
 </networks>
```

> Information about CI checks: https://clickhouse.tech/docs/en/development/continuous-integration/

### CODE REVIEW

**Comment by nikitamikhaylov** [comment]
@Mergifyio update

File: src/Common/tests/gtest_helper_functions.h [hunk 1]
```diff
@@ -0,0 +1,59 @@
+#pragma once
+
+#include "Poco/DOM/Document.h"
+#include "Poco/DOM/NodeList.h"
+#include "Poco/DOM/NamedNodeMap.h"
+
+
+inline char *getTempFileWithContents(const char *fileContents)
```
Comment (nikitamikhaylov) [inline src/Common/tests/gtest_helper_functions.h:8]: Consider to use `DB::WriteBufferFromFile` with path starting with `/tmp/...`

File: src/Common/tests/gtest_helper_functions.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+{
+    auto *file_pointer = std::tmpnam (nullptr);
+    auto *file_stream = std::fopen(file_pointer, "w+");
+    std::fputs(fileContents, file_stream);
+    std::fclose(file_stream);
+
+    return file_pointer;
+}
+
+inline std::string xmlNodeAsString(Poco::XML::Node* &pNode)
```
Comment (nikitamikhaylov) [inline src/Common/tests/gtest_helper_functions.h:18]: Reference to pointer?

File: src/Common/tests/gtest_helper_functions.h (hunk 1 above)
Comment (mdonkers) [reply src/Common/tests/gtest_helper_functions.h:8]: Do we need to take Windows builds into account or not needed? As on Windows the `/tmp/...` path wouldn't be possible.
Can also store files in some other place, but then the question is where? (preferably they get cleaned automatically if `remove` fails for whatever reason, and must be possible to consistently reference them from tests, which was failing with the lookup before)

File: src/Common/tests/gtest_helper_functions.h (hunk 2 above)
Comment (mdonkers) [reply src/Common/tests/gtest_helper_functions.h:18]: Had to look it up and found some example I (partially) copy-pasted and had to further modify to work correctly.
But I'll change the parameter.

File: src/Common/tests/gtest_helper_functions.h (hunk 1 above)
Comment (mdonkers) [reply src/Common/tests/gtest_helper_functions.h:8]: @nikitamikhaylov any idea how we can further improve this unit test? I believe external files are only used for functional / integration tests, nowhere else.
Initially I thought about using `YAMLParser::processNode` directly, but this is currently not exposed and I didn't want to change the interface for this class.
gtest apparently has something like `FRIEND_TEST` to expose internal functions just for testing. Not sure if we can / should use that instead?

Comment (alesapin) [reply src/Common/tests/gtest_helper_functions.h]: No, ClickHouse doesn't build for Windows.

**Review by nikitamikhaylov — APPROVED** [review]

**Comment by vitlibar** [comment]
I don't agree with that change. I think it actually made the conversion from YAML to XML less conventional and less expectable. I checked many popular conversion tools (for example, see https://www.browserling.com/tools/yaml-to-xml, but there are other online tools and python too) and all of them converted

```
networks:
  - ip: "10.1.6.168"
  - ip: "::1"
  - ip: "127.0.0.1"
```

to

```
<networks>
    <ip>10.1.6.168</ip>
</networks>
<networks>
    <ip>::1</ip>
</networks>
<networks>
    <ip>127.0.0.1</ip>
</networks>
```

And if you want to get this XML

```
<networks>
    <ip>10.1.6.168</ip>
    <ip>::1</ip>
    <ip>127.0.0.1</ip>
</networks>
```

with those tools you actually need to use the following YAML:

```
networks:
    ip:
      - 10.1.6.168
      - ::1
      - 127.0.0.1
```

And also other people don't understand why our conversion work the way you did it in this PR (see https://github.com/ClickHouse/ClickHouse/issues/40561#issuecomment-1235191114)

So I think it would be better to come back the previous behaviour, see https://github.com/ClickHouse/ClickHouse/pull/41044

**Comment by mdonkers** [comment]
Okay, I see the issue and we can change also on our end.
It comes from the way how we create the YAML in Golang, but can construct it in a different way: https://github.com/ClickHouse/data-plane-application/pull/1689
Though personally I find it less logical, we should stick with conventions.

**Comment by vitlibar** [comment]
> personally I find it less logical, but we should stick with conventions.

Perhaps. I actually tried to understand why the conventions are as they are. It seems it's because YAML can be considered as a superset of JSON, for example

```
seq:
  - k1: val1
    k2: val2
  - k3: val3
```

is equivalent to

```
{
    "seq": [
        {
            "k1": "val1",
            "k2": "val2"
        },
        {
            "k3": "val3"
        }
    ]
}
```

and then they repeat `seq` in xml because they need to represent the fact that `k1`, `k2` and `k3` are in two different dictionaries:

```
<seq>
    <k1>val1</k1>
    <k2>val2</k2>
</seq>
<seq>
    <k3>val3</k3>
</seq>
```

**Pull request merged**